### PR TITLE
MM-41532 - Gekidou Snack Bars

### PR DESCRIPTION
**Pull request by avinashlng1080**
#### Summary
This PR adds the Toasts component after each post option menu action; for e.g. we're now going to show a toast after a link/message is copied to the clipboard.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-41532

#### Device Information
This PR was tested on: 

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: app/screens/snack_bar/constants.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        backgroundColor: 'centerChannelColor',
+        canUndo: true,
+    },
+    FAILED_TO_SAVE_MESSAGE: {
+        id: 'snack.bar.image.save.failed',
+        defaultMessage: 'Failed to save image',
+        iconName: 'alert-outline',
+        backgroundColor: 'dndIndicator',
+        canUndo: false,
+    },
```
Comment (avinashlng1080) [inline app/screens/snack_bar/constants.ts:49]: to add network disconnected type in this object

File: app/screens/snack_bar/useToastToggler.tsx [hunk 2]
```diff
@@ -0,0 +1,22 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {useEffect, useState} from 'react';
+import {Navigation} from 'react-native-navigation';
+
+const useToastToggler = (componentId: string, isDismissible: boolean) => {
```
Comment (avinashlng1080) [inline app/screens/snack_bar/useToastToggler.tsx:7]: Might not be needed... I will implement the toasts for the network connectivity as well and potentially re-used it...else I will remove it and put it back in the Snackbar screen

File: app/screens/index.tsx [hunk 3]
```diff
@@ -136,6 +136,11 @@ Navigation.setLazyComponentRegistrator((screenName) => {
         case Screens.THREAD:
             screen = withServerDatabase(require('@screens/thread').default);
             break;
+        case Screens.SNACK_BAR: {
+            const snackBarScreen = withIntl(require('@screens/snack_bar').default);
+            Navigation.registerComponent(Screens.SNACK_BAR, () => snackBarScreen);
+            break;
+        }
```
Comment (enahum) [inline app/screens/index.tsx:143]:
```suggestion
        case Screens.SNACK_BAR: {
            const snackBarScreen = withServerDatabase(require('@screens/snack_bar').default);
            Navigation.registerComponent(Screens.IN_APP_NOTIFICATION, () =>
                Platform.select({
                    default: snackBarScreen,
                    ios: withSafeAreaInsets(snackBarScreen),
                }),
            );
            break;
        }
```

File: app/screens/post_options/options/copy_permalink_option/copy_permalink_option.tsx [hunk 4]
```diff
@@ -24,6 +26,7 @@ const CopyPermalinkOption = ({teamName, post}: Props) => {
         const permalink = `${serverUrl}/${teamName}/pl/${post.id}`;
         Clipboard.setString(permalink);
         dismissBottomSheet(Screens.POST_OPTIONS);
+        showToast(SNACK_BAR_TYPE.LINK_COPIED);
```
Comment (avinashlng1080) [inline app/screens/post_options/options/copy_permalink_option/copy_permalink_option.tsx:29]:
```suggestion
        showSnackBar(SNACK_BAR_TYPE.LINK_COPIED);
```

File: app/screens/post_options/options/copy_permalink_option/copy_permalink_option.tsx [hunk 5]
```diff
@@ -8,6 +8,8 @@ import {Screens} from '@constants';
 import {useServerUrl} from '@context/server';
 import {t} from '@i18n';
 import {dismissBottomSheet} from '@screens/navigation';
+import {SNACK_BAR_TYPE} from '@screens/snack_bar/constants';
+import {showToast} from '@utils/toast';
```
Comment (avinashlng1080) [inline app/screens/post_options/options/copy_permalink_option/copy_permalink_option.tsx:12]:
```suggestion
import {showToast} from '@utils/snack_bar';
```

File: app/screens/post_options/options/copy_permalink_option/copy_permalink_option.tsx [hunk 6]
```diff
@@ -8,6 +8,8 @@ import {Screens} from '@constants';
 import {useServerUrl} from '@context/server';
 import {t} from '@i18n';
 import {dismissBottomSheet} from '@screens/navigation';
+import {SNACK_BAR_TYPE} from '@screens/snack_bar/constants';
```
Comment (avinashlng1080) [inline app/screens/post_options/options/copy_permalink_option/copy_permalink_option.tsx:11]: move this to the constants folder

File: app/screens/post_options/options/follow_option.tsx [hunk 7]
```diff
@@ -40,8 +42,15 @@ const FollowThreadOption = ({thread}: FollowThreadOptionProps) => {
         }
     }
 
+    const unFollowThread = useCallback(() => {
+        // eslint-disable-next-line no-console
```
Comment (avinashlng1080) [inline app/screens/post_options/options/follow_option.tsx:46]: Add a ticket

File: app/screens/post_options/options/follow_option.tsx [hunk 8]
```diff
@@ -40,8 +42,15 @@ const FollowThreadOption = ({thread}: FollowThreadOptionProps) => {
         }
     }
 
+    const unFollowThread = useCallback(() => {
```
Comment (avinashlng1080) [inline app/screens/post_options/options/follow_option.tsx:45]: @matthewbirtch : When we follow a thread, the snackbar shows up with the `undo` action.  But now, on long pressing on the thread, we have the `unfollow` option.  Should we be displaying the snackbar again if the user presses on the `unfollow` button?

File: app/screens/post_options/options/follow_option.tsx [hunk 9]
```diff
@@ -40,8 +42,15 @@ const FollowThreadOption = ({thread}: FollowThreadOptionProps) => {
...
+        // eslint-disable-next-line no-console
+        console.log('TO IMPLEMENT UNFOLLOW THREAD');
+    }, []);
+
     const handleToggleFollow = () => {
         //todo:
+
+        showToast(SNACK_BAR_TYPE.FOLLOW_THREAD, unFollowThread);
```
Comment (avinashlng1080) [inline app/screens/post_options/options/follow_option.tsx:53]: Add an if/else structure to align with the 'toggle' functionality.

File: app/screens/post_options/options/follow_option.tsx (hunk 8 above)
Comment (avinashlng1080) [inline app/screens/post_options/options/follow_option.tsx:45]:
```suggestion
    const undo = useCallback(() => {
```

also follows the toggle functionality

File: app/screens/snack_bar/constants.ts [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    LINK_COPIED: null,
+    MESSAGE_COPIED: null,
+    FOLLOW_THREAD: null,
+    MUTE_CHANNEL: null,
+    FAILED_TO_SAVE_MESSAGE: null,
+});
+
+export const SNACK_BAR_CONFIG = {
+    LINK_COPIED: {
+        id: 'snack.bar.link.copied',
```
Comment (avinashlng1080) [inline app/screens/snack_bar/constants.ts:16]: wrap with t function

File: app/screens/snack_bar/constants.ts [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+        defaultMessage: 'Message copied to clipboard',
+        iconName: 'content-copy',
+        backgroundColor: 'centerChannelColor',
+        canUndo: false,
+    },
+    FOLLOW_THREAD: {
+        id: 'snack.bar.follow.thread',
+        defaultMessage: 'You\'re now following this thread',
+        iconName: 'message-check-outline',
+        backgroundColor: 'centerChannelColor',
```
Comment (avinashlng1080) [inline app/screens/snack_bar/constants.ts:33]:
```suggestion
       themeColor: 'centerChannelColor',
```

**Comment by Willyfrog** [comment]
@avinashlng1080 is this ready for review? or is it still a draft?

File: app/screens/post_options/options/follow_option.tsx (hunk 8 above)
Comment (matthewbirtch) [reply app/screens/post_options/options/follow_option.tsx]: @avinashlng1080 I don't think we should show the snackbar when you follow a thread - only when you `unfollow` a thread.

Comment (enahum) [reply app/screens/post_options/options/follow_option.tsx]: I would say that they snackbar is not needed for either one as there is UI changes in the channel or thread screen already

**Comment by avinashlng1080** [comment]
@Willyfrog  The work on this ticket is done.  However, given that it is a Beta ticket, we have parked it for now.  Once we revisit it, we can include the new UI considerations/changes ( if any ) into it.  

@matthewbirtch  FYI.

Link to the conversation: https://community.mattermost.com/core/pl/os6x1ging78adff15d3eajkicw

**Pull request closed without merging**